render/detailed/links.go

```diff
@@ -0,0 +1,167 @@
+			docker.CPUTotalUsage: prepareTemplate(`namespace_name:container_cpu_usage_seconds_total:sum_rate{name="{{.Label}}"}`),
+		},
+		report.Service: {
+			docker.MemoryUsage:   prepareTemplate(`namespace_name:container_memory_usage_bytes:sum{name="{{.Label}}",monitor=""}`),
+			docker.CPUTotalUsage: prepareTemplate(`namespace_name:container_cpu_usage_seconds_total:sum_rate{name="{{.Label}}"}`),
+		},
+	}
+)
+
+// SetMetricsGraphURL sets the URL we deduce our eventual metric link from.
+// Supports placeholders such as `:orgID` and `:query`. An empty url disables
+// this feature. If the `:query` part is missing, a JSON version will be
+// appended, see `queryParamsAsJSON()` for more info.
+func SetMetricsGraphURL(url string) {
+	metricsGraphURL = url
+}
+
+// NodeLinks returns the links of a node. The links are collected
+// by a predefined set but filtered depending on whether a query
+// is configured or not for the particular topology.
+func NodeMetricLinks(_ report.Report, n report.Node) []MetricLink {
+	if metricsGraphURL == "" {
+		return nil
+	}
+
+	queries := topologyQueries[n.Topology]
+	if len(queries) == 0 {
+		return nil
+	}
+
+	links := []MetricLink{}
+	for _, link := range linkTemplates {
+		if _, ok := queries[link.ID]; ok {
+			links = append(links, link)
+		}
+	}
+
+	return links
+}
+
+// RenderLinks executes the templated links by supplying the node summary as data.
+// It returns the modified summary.
+func RenderMetricLinks(summary NodeSummary, n report.Node) NodeSummary {
+	queries := topologyQueries[n.Topology]
+	if len(queries) == 0 || len(summary.MetricLinks) == 0 {
+		return summary
+	}
+
+	links := []MetricLink{}
+	var bs bytes.Buffer
+	for _, link := range summary.MetricLinks {
+		tpl := queries[link.ID]
+		if tpl == nil {
+			continue
+		}
+
+		bs.Reset()
+		if err := tpl.Execute(&bs, summary); err != nil {
+			continue
+		}
+
+		link.URL = buildURL(bs.String())
+		links = append(links, link)
+	}
+	summary.MetricLinks = links
+
+	return summary
+}
+
+// buildURL puts together the URL by looking at the configured
+// `metricsGraphURL`.
+func buildURL(query string) string {
+	if strings.Contains(metricsGraphURL, urlQueryVarName) {
+		return strings.Replace(metricsGraphURL, urlQueryVarName, url.PathEscape(query), -1)
+	}
+
+	params, err := queryParamsAsJSON(query)
+	if err != nil {
+		return ""
+	}
+
+	if metricsGraphURL[len(metricsGraphURL)-1] != '/' {
+		metricsGraphURL += "/"
+	}
+
+	return metricsGraphURL + url.PathEscape(params)
+}
+
+// queryParamsAsJSON packs the query into a JSON of the
+// format `{"cells":[{"queries":[$query]}]}`.
+func queryParamsAsJSON(query string) (string, error) {
+	type cell struct {
+		Queries []string `json:"queries"`
+	}
+	type queryParams struct {
+		Cells []cell `json:"cells"`
+	}
+	params := &queryParams{[]cell{{[]string{query}}}}
+
+	bs, err := json.Marshal(params)
```

Comment:
We use the ugorji library for marshaling in most places (if not all) since it's faster.